concepts/0003-protocols/semver.md

```diff
@@ -1,93 +1,51 @@
-correctly by agents.
-
-### Version Negotiation
-
-The semver portion of a [message type or protocol identifier URI](uris.md) is
-used to establish the version of a protocol that parties use during an
-interaction.
-
-#### Initiator
-
-Unless Alice's agent (the initiator of a protocol) knows from prior history
-that it should do something different, it should begin a protocol using the
-highest version number that it supports. For example, if A.1
-supports versions 2.0 through 2.2 of protocol X, it should use 2.2 as the
-version in the message type of its first message.
-
-#### Recipient Rules
-
-Agents for Bob (the recipient) should reject messages from protocols with major
-versions different from those they support. For major version 0, they should also
-reject protocols with minor versions they don't support, since semver stipulates
-that [features are not stable before 1.0](https://semver.org/#spec-item-4). For
-example, if B.1 supports only versions 2.0 and 2.1 of protocol X, it should reject
-any messages from version 3 or version 1 or 0. In most cases, rejecting a message
-means sending a `problem-report` that the message is unsupported. The `code` field
-in such messages should be `version-not-supported`. Agents that receive such a
-`problem-report` can then use the [Discover Features Protocol](
-../../features/0031-discover-features/README.md) to resolve version problems.
-
-Recipient agents should accept messages that differ from their own supported version
-of a protocol only in the patch, prerelease, and/or build fields, whether these
-differences make the message earlier or later than the version the recipient prefers.
-These messages will be robustly compatible.
-
-For major version >= 1, recipients should also accept messages that differ only in that
-the message's minor version is earlier than their own preference. In such a case, the
-recipient should degrade gracefully to use the earlier version of the protocol. If the
-earlier version lacks important features, the recipient may optionally choose to send,
-in addition to a response, a `problem-report` with code `version-with-degraded-features`.
-
-If a recipient supports protocol X version 1.0, it should tentatively
-accept messages with later minor versions (e.g., 1.2). Message types that
-differ in only in minor version are guaranteed to be compatible *for the
-feature set of the earlier version*. That is, a 1.0-capable agent can support
-1.0 features using a 1.2 message, though of course it will lose any features
-that 1.2 added. Thus, accepting such a message could have two possible outcomes:
-
-1. The message at version 1.2 might look and behave exactly like it did at version
-1.0, in which case the message will process without any trouble.
+Within a given major version of a protocol, an agent should:
 
-2. The message might contain some fields that are unrecognized and need to be ignored.
+- specify a minimum minor version supported, defaulting to "0"
```

Comment:
How does it specify this? This simplification loses a lot of important detail. It is not equivalent; it's rather a substantive change to agreed upon rules.

Reply:
How is that a change?  I had made the statement that "all minor versions of a protocol must be supported", and someone (I thought you) had objected that an agent may refuse to support a minor version below a given number.  That is what is being conveyed.

In ACA-Py, we are putting into code:

- each major version of each protocol and for each:
    - the minimum minor version allowed
    - the supported/default minor version
    - a link to the commit in github of the RFC for the protocol major version.

All of that is our way to specify that information.  Others of course can do something different or nothing.

Changed the wording to "respond to" from "specify" in this and the next line.

Reply:
When I read the word "specify" I assumed you were talking about some behavior in the protocols themselves -- that something in messages sent by an agent would specify which versions it supported. We have no standard way of doing this, so when I read it that way, I got concerned. If all you meant by "specify" was "decide and document" then I am cool with that word.

Regarding substantive change: the new and simplified text changes the intent of this section from being about which version of a protocol is used, to being about which version of a protocol elicits a compatible response. Those are not the same thing. The effect of the original text would be to advocate the following behavior:

>A sends v 1.3 of a message to B. B, being more up-to-date, prefers version 1.5. However, B responds, and the response uses version 1.3, because the protocol version has been negotiated down to the least common denominator. Both A and B know and explicitly acknowledge that no features above 1.3 will take effect in the interaction.

The updated text allows the following:

>A sends v 1.3 of a message to B. B, being more up-to-date, prefers version 1.5. However, B responds. The response uses version 1.5. Now A has to decide whether to support 1.5 or not. The protocol version is still unclear; are they both speaking the same version of the protocol? Is B aware that any features in versions >1.3 are being dropped, or were they just careless about inspecting versions?